python/pyproject.toml

```diff
@@ -61,6 +61,7 @@ thrift = { version = ">=0.13.0,<1.0.0", optional = true }
 boto3 = { version = ">=1.17.106", optional = true }
 s3fs = { version = ">=2021.08.0,<2024.1.0", optional = true } # Upper bound set arbitrarily, to be reassessed in early 2024.
 adlfs = { version = ">=2021.07.0,<2024.1.0", optional = true } # Upper bound set arbitrarily, to be reassessed in early 2024.
+psycopg2-binary = { version = ">=2.9.6", optional = true }
```

Comment:
Would it make sense to use a library like SQLAlchemy to avoid direct dependencies on database drivers? I'm assuming that there's a good way to manage dependencies and keep SQL generic provided by that library.

Reply:
I've also raised this with Fokko.  I'm not a Python expert and from what I know, there is the [DB-API spec](https://peps.python.org/pep-0249/).  But I find the interface is not as good as JDBC's, especially when it comes to SQL statement parameters.  Maybe my limited experience with it is clouding my judgement, so feel free to correct me.  I'm all for supporting "any" mainstream DB and I think that's what pyiceberg users are also expecting (at least I am!)  Now, is SQLAlchemy the only option, I'm not sure.

Reply:
Yes, so there are multiple things here. The `DB-API` spec should be able to abstract the different databases, while SQLAlchemy is a full ORM layer.

Personally, I'm not a huge fan of ORM layers in general because it takes the common denominator, and you won't be able to optimize (SQLAlchemy allows you to do this to some extent). Also, with Airflow we experienced some memory pressure, but I won't expect PyIceberg to have many open and concurrent connections.

I think @cccs-eric also noticed that while libraries implement the `DB_API` specification, there are still subtle differences:

- [Postgres prepared statements](https://www.psycopg.org/docs/usage.html): `INSERT INTO some_table (an_int, a_date, a_string) VALUES (%s, %s, %s);`.
- [SQLite prepared statements](https://docs.python.org/3/library/sqlite3.html): `INSERT INTO movie VALUES(?, ?, ?)`. 

I don't think not-using prepared statements is an option due to security considerations. Therefore I would lean to switching to `SQLAlchemy`. Another nice feature is that it also supports [generating the `CREATE TABLE` statements](https://www.tutorialspoint.com/sqlalchemy/sqlalchemy_core_creating_table.htm). This also makes it easier to add engines, and run tests against them.

Reply:
@Fokko Now that we have SQLAlchemy in place, how does it work with driver dependencies?

Reply:
What do you think of:

```
pip3 install pyiceberg[sql-postgres]
```

This will install sqlalchemy, together with the Postgres driver.